### PR TITLE
Fix: Avoid infinite recursion when source and destinations are same while using `cp -R`

### PR DESCRIPTION
**Pull request by Narasimha1997**
Proposing a solution for #3016 
The infinite recursion can be prevented when running `cp -r <source_dir> <destination_dir> by handling these two cases:

### 1. Source and Destination refer to same canonicalized path, after resolving all symbolic links etc
```
# Example: 
cp -r /home/ubuntu/parent   /home/ubuntu/parent 
```
### 2. Canonicalized destination has a canonicalized source as prefix:
This can also lead to recursion, imagine there is a directory called `child` under `/home/ubuntu/parent` (i.e `/home/ubuntu/parent/child`, then the following example leads to recursion:
```
cp -r /home/ubuntu/parent /home/ubuntu/parent/child
```
The official GNU coreutil's `cp` handles both of these cases.

The above two cases can be solved by simply checking for case 2, i.e don't allow copy if destination has source in it's prefix.
I added a function:
```rust
pub fn path_has_prefix(p1: &Path, p2: &Path) -> io::Result<bool> {
    let pathbuf1 = canonicalize(p1, MissingHandling::Normal, ResolveMode::Logical)?;
    let pathbuf2 = canonicalize(p2, MissingHandling::Normal, ResolveMode::Logical)?;

    Ok(pathbuf1.starts_with(pathbuf2))
}
```
Which checks that and calling it in `copy_directory` function.

### CODE REVIEW

**Comment by sylvestre** [comment]
Fun, could you please add a test to make sure we don't regress?

**Comment by Narasimha1997** [comment]
@sylvestre added tests as you suggested.

File: tests/by-util/test_cp.rs [hunk 1]
```diff
@@ -1444,3 +1444,48 @@ fn test_cp_archive_on_nonexistent_file() {
...
+        .arg("parent1")
+        .arg("parent1")
+        .fails()
+        .stderr_contains("cannot copy a directory");
+    // case-2: copy parent1 -> parent1/child should fail
+    scene
+        .ucmd()
+        .arg("-R")
+        .arg("parent1")
+        .arg("parent1")
```
Comment (shoriminimoe) [inline tests/by-util/test_cp.rs:1473]: I believe the second argument here should be `parent1/child` based on the comment.

Comment (Narasimha1997) [reply tests/by-util/test_cp.rs]: My bad! Fixed this.

**Comment by Narasimha1997** [comment]
@tertsdiepraam some merge issues with this branch

**Comment by tertsdiepraam** [comment]
Can you try to fix them? I think it's just that a test was added at the end of the file in both branches and that we should keep both.

**Pull request merged**